server/docs/design/block-verification.md

```diff
@@ -0,0 +1,198 @@
+7. [Enums](#enums)
+8. [Metrics](#metrics)
+9. [Exceptions](#exceptions)
+
+## Purpose
+
+The purpose of the Block Verification feature is to ensure that blocks received from consensus nodes are valid and have not been tampered with. This is achieved by re-calculating the block hash and verifying it against the signature provided by the consensus node.
+
+## Goals
+
+1) The block-node must re-create the block hash from the block items and verify that it matches the hash implied by the signature.
+2) If verification fails, the block should be considered invalid, and appropriate error-handling procedures must be triggered.
+
+
+## Terms
+
+- Consensus Node (CN): A node that produces and provides blocks.
+- Block Items: The block data pieces (header, events, transactions, transaction result, state changes, proof) that make up a block.
+- Block Hash: A cryptographic hash representing the block’s integrity.
+- Signature: The cryptographic signature of the block hash created by Network private key (aka: LedgerId)
+- Public Key: The public key of the LedgerId that signed the block.
+
+
+## Entities
+
+- **VerificationHandler** 
+  - Receives the stream of block items in the form of List<BlockItemsUnparsed> from the unparsed and unverified block items ring buffer. 
+  - When it detects a block_header, it creates a BlockHashingSession using the BlockHashingSessionFactory, providing it with the initial block items (and internally, the session will handle asynchronous hashing).
+  - Adds subsequent block items to the session, including the block_proof.
+  - Once the block_proof is received, calls completeHashing() on the BlockHashingSession, then releases the thread back to the UnverifiedRingBuffer.
+  - Does not block waiting for verification; the hash computation and verification continue asynchronously.
+- **BlockHashingSessionFactory**
+  - Creates new BlockHashingSession instances, provides them with a ExecutorService.
+- **BlockHashingSession**
+  - Holds all necessary block data.
+  - Accepts block items incrementally. (continues to compute them asynchronously)
+  - Once the block_proof is provided, finalizes the hash computation asynchronously.
+  - After computing the final hash, calls SignatureVerifier for verification.
+- **SignatureVerifier**
+  - Verifies the signature by comparing the computed hash to the hash implied by the signature (using the public key).
+  - Calls BlockStatusManager with the verification result.
+- **BlockStatusManager**
+  - Receives verification results from SignatureVerifier.
+  - Updates block status and triggers any necessary recovery or follow-up processes depending on the outcome.
+
+## Design
+
+1. The `VerificationHandler` receives the list of block items from the unverified ring buffer.
+2. When the block_header is detected, the `VerificationHandler` creates a `BlockHashingSession` using the `BlockHashingSessionFactory`.
+3. The `BlockHashingSession` accepts subsequent block items incrementally.
```

Comment:
Just a note, there's no need to keep the numbers up, they can all be `1.`.  Markdown translates to an ordered list, and the renderer (browser) generates all the numbers.

Reply:
✅ Done!